convert2rhel/unit_tests/conftest.py

```diff
@@ -19,6 +20,39 @@ def is_py2():
     return sys.version_info[:2] <= (2, 7)
 
 
+@pytest.fixture
+def clear_loggers():
+    """
+    Remove handlers from the loggers.
+
+    If logger handlers are created while we are using capsys or capfd, then
+    the logger handlers get created with pytest's capture.  If pytest's
+    capture feature is disabled later, the logger handlers will get errors
+    trying to write to a closed file.  This is this issue:
+
+        https://github.com/pytest-dev/pytest/issues/5502
+
+    In our unittests, there are some tests that use pexpect that have to
+    disable pytest's capture otherwise pexpect-2.3 will malfunction.  The
+    combination of these two things causes logging calls during the utils
+    tests to complain that they are trying to log to a closed file handle.
+
+    Adding this fixture to any unit test which both uses capsys/capfd and
+    creates the logger handlers (by calling
+    `convert2rhel.logger.setup_logger_handler()`) fixes the issue.  The
+    fixture works by removing the handlers that `setup_logger_handler()`
+    (actually, all of the handlers) created when the test completes.
+    """
+    # Nothing to do in setup
+    yield
+    # In teardown, we clear all the logger handlers.
+    loggers = [logging.getLogger()] + list(logging.Logger.manager.loggerDict.values())
+    for logger in loggers:
+        handlers = getattr(logger, "handlers", [])
+        for handler in handlers:
+            logger.removeHandler(handler)
```

Comment:
```suggestion
        if not hasattr(logger, "handlers"):
            continue
        for handler in logger.handlers[:]:
            logger.removeHandler(handler)
```
This for loop does not work correctly because by calling the `removeHandler()` you effectively mutate the list you're iterating on, returned by `getattr(logger, "handlers", [])`. The end result is that not all handlers are removed from the logger. In my testing, just two out of four handlers are removed. Is this the reason for duplicating the code from this fixture in the tests themselves? Because iterating over the handlers the second time allowed to finally remove the rest of them.

What fixes the fixture is making a copy of the handlers list, for example by using a slice notation (inspiration: https://youtu.be/xeLecww65Zg?t=805).

By the way, the credit goes to this person who figured this out: https://stackoverflow.com/a/7484605/3499937).